src/library/scala/collection/immutable/VectorMap.scala

```diff
@@ -0,0 +1,150 @@
+  *  of using extra memory and generally lower performance for other operations
+  *
+  *  @tparam K      the type of the keys contained in this vector map.
+  *  @tparam V      the type of the values associated with the keys in this vector map.
+  *
+  * @author Matthew de Detrich
+  * @version 2.13
+  * @since 2.13
+  * @define coll immutable vector map
+  * @define Coll `immutable.VectorMap`
+  */
+final class VectorMap[K, +V] private[immutable] (
+    private val fields: Vector[K],
+    private val underlying: Map[K, (Int, V)])
+    extends AbstractMap[K, V]
+    with SeqMap[K, V]
+    with MapOps[K, V, VectorMap, VectorMap[K, V]]
+    with StrictOptimizedIterableOps[(K, V), Iterable, VectorMap[K, V]] {
+
+  override protected[this] def className: String = "VectorMap"
+
+  def updated[V1 >: V](key: K, value: V1): VectorMap[K, V1] = {
+    underlying.get(key) match {
+      case Some(oldIndexWithValue) =>
+        new VectorMap(fields,
+          underlying.updated(key, (oldIndexWithValue._1, value)))
+      case None =>
+        new VectorMap(
+          fields :+ key,
+          underlying.updated(key, (fields.length + 1, value)))
+    }
+  }
+
+  override def withDefault[V1 >: V](d: K => V1): Map.WithDefault[K, V1] =
+    new Map.WithDefault(this, d)
+
+  override def withDefaultValue[V1 >: V](d: V1): Map.WithDefault[K, V1] = new Map.WithDefault[K, V1](this, _ => d)
+
+  def iterator: Iterator[(K, V)] = new Iterator[(K, V)] {
+    private val fieldsIterator = fields.iterator
+
+    override def hasNext: Boolean = fieldsIterator.hasNext
+
+    override def next(): (K, V) = {
+      val field = fieldsIterator.next()
+      (field, underlying(field)._2)
+    }
+  }
+
+  def get(key: K): Option[V] = underlying.get(key) match {
+    case Some(v) => Some(v._2)
+    case None => None
+  }
+
+  def remove(key: K): VectorMap[K, V] = {
+    underlying.get(key) match {
+      case Some((index, _)) =>
+        new VectorMap(fields.patch(index, Nil, 1), underlying - key)
+      case _ =>
+        this
+    }
+  }
+
+  override def mapFactory: MapFactory[VectorMap] = VectorMap
+
+  override def size: Int = fields.size
+
+  override def knownSize: Int = fields.size
+
+  override def isEmpty: Boolean = fields.isEmpty
+
+  override final def contains(key: K): Boolean = underlying.contains(key)
+
+  override def head: (K, V) = iterator.next()
+
+  override def last: (K, V) = {
+    val last = fields.last
+    (last, underlying(last)._2)
+  }
+
+  override def lastOption: Option[(K, V)] = {
+    fields.lastOption match {
+      case Some(last) => Some(last, underlying(last)._2)
+      case None => None
+    }
+  }
+
+  override def tail: VectorMap[K, V] = {
+    new VectorMap(fields.tail, underlying.remove(fields.last))
+  }
+
+  override def init: VectorMap[K, V] = {
+    new VectorMap(fields.init, underlying.remove(fields.head))
+  }
+
+  // Only care about content, not ordering for equality
+  override def equals(that: Any): Boolean =
```

Comment:
What do you think, is there a need for an equals check which cares about the ordering?

Reply:
We can not do that since `VectorMap` extends `Map`, so it must conform to `Map` equality:

~~~ scala
(vectorMap1 == map && vectorMap2 == map) implies (vectorMap1 == vectorMap2)
~~~

Because equality must be transitive, if we have two `VectorMap`s that are equal to a `Map` (without taking elements order into account), then these two `VectorMap` must be equal (thus we can not take elements order into account, otherwise the two `VectorMap`s might not be equal).

Reply:
Makes sense, so it can be only a new method, but not sure how this would align with the collection architecture.

Reply:
Or you can just write `vectorMap1.toSeq == vectorMap2.toSeq`, which takes element order into account.

Reply:
or even `(vm1 corresponds vm2)(_ == _)`.

Reply:
could even do `sameElements` I think